src/app/banking/vouchers/buy-vouchers.component.ts

```diff
@@ -0,0 +1,186 @@
+import { FormGroup, FormControl, Validators } from '@angular/forms';
+import { ConfirmationMode } from 'app/shared/confirmation-mode';
+import { validateBeforeSubmit } from 'app/shared/helper';
+
+export type BuyVouchersStep = 'select-type' | 'form' | 'confirm';
+
+/**
+ * Component used to buy vouchers.
+ * The data requested for the first time is to get all possible voucher types.
+ */
+@Component({
+  selector: 'buy-vouchers',
+  templateUrl: 'buy-vouchers.component.html',
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class BuyVouchersComponent extends BasePageComponent<VoucherDataForBuy>
+  implements OnInit {
+
+  ConfirmationMode = ConfirmationMode;
+
+  step$ = new BehaviorSubject<BuyVouchersStep>(null);
+
+  singleType = false;
+  user: string;
+  form: FormGroup;
+
+  confirmationPassword: FormControl;
+  confirmationMode$ = new BehaviorSubject<ConfirmationMode>(null);
+
+  customFieldControlsMap: Map<string, FormControl>;
+
+  // The data for a specific voucher type
+  dataTypeForBuy: VoucherDataForBuy;
+  canConfirm: boolean;
+
+  constructor(injector: Injector, private voucherService: VouchersService) {
+    super(injector);
+  }
+
+  get step(): BuyVouchersStep {
+    return this.step$.value;
+  }
+  set step(step: BuyVouchersStep) {
+    this.step$.next(step);
+  }
+
+  ngOnInit() {
+    super.ngOnInit();
+    const params = this.route.snapshot.params;
+    this.user = this.authHelper.isSelf(params.user)
+      ? this.ApiHelper.SELF
+      : params.user;
+
+    this.addSub(
+      this.voucherService.getVoucherDataForBuy({ user: this.user })
+        .subscribe(data => this.data = data)
+    );
+  }
+
+  onDataInitialized(data: VoucherDataForBuy) {
+    const types = data.types || [];
+    if (types.length === 1) {
+      this.singleType = true;
+      this.toForm(types[0]);
+    } else {
+      this.step = 'select-type';
+    }
+  }
+
+  get currency(): Currency {
+    return this.dataTypeForBuy.account.currency;
+  }
+
+  /**
+   * Final action: buy the vouchers
+   */
+  buy(confirmationPassword?: string) {
+    if (confirmationPassword) {
+      this.confirmationPassword.setValue(confirmationPassword);
+    }
+
+    if (this.confirmationPasswordInput && !validateBeforeSubmit(this.confirmationPassword)) {
+      return;
+    }
+
+    const body = this.form.value;
+    body.type = this.dataTypeForBuy.type.id;
+    const params = {
+      user: this.user,
+      confirmationPassword: this.confirmationPasswordInput ? this.confirmationPassword.value : null,
+      body: body
+    };
+    this.addSub(
+      this.voucherService.buyVouchers(params)
+        .subscribe((ids: string[]) => {
+          if (ids.length === 1) {
+            this.router.navigate(['banking', 'vouchers', ids[0]]);
+          } else {
+            this.router.navigate(['banking', this.user, 'vouchers', 'bought']);
+          }
+          this.notification.info(this.i18n.voucher.buy.done);
```

Comment:
I think a snackbar would be better, but that's debatable.

Reply:
I agree